internal/controller/config.go

```diff
@@ -0,0 +1,121 @@
+
+import (
+	"flag"
+	"fmt"
+	"runtime"
+	"time"
+
+	log "github.com/sirupsen/logrus"
+	"go.opentelemetry.io/ebpf-profiler/tracer"
+)
+
+type Config struct {
+	BpfVerifierLogLevel    uint
+	CollAgentAddr          string
+	Copyright              bool
+	DisableTLS             bool
+	MapScaleFactor         uint
+	MonitorInterval        time.Duration
+	ClockSyncInterval      time.Duration
+	NoKernelVersionCheck   bool
+	PprofAddr              string
+	ProbabilisticInterval  time.Duration
+	ProbabilisticThreshold uint
+	ReporterInterval       time.Duration
+	SamplesPerSecond       int
+	SendErrorFrames        bool
+	Tracers                string
+	VerboseMode            bool
+	Version                bool
+
+	Fs *flag.FlagSet
+}
+
+const (
+	exitParseError int = 2
+
+	// 1TB of executable address space
+	MaxArgMapScaleFactor = 8
+)
+
+// Dump visits all flag sets, and dumps them all to debug
+// Used for verbose mode logging.
+func (cfg *Config) Dump() {
+	log.Debug("Config:")
+	cfg.Fs.VisitAll(func(f *flag.Flag) {
+		log.Debug(fmt.Sprintf("%s: %v", f.Name, f.Value))
+	})
+}
+
+// Validate runs validations on the provided configuration, and returns errors
+// if invalid values were provided.
+func (cfg *Config) Validate() error {
+	if cfg.SamplesPerSecond < 1 {
+		return parseError("invalid sampling frequency: %d", cfg.SamplesPerSecond)
+	}
+
+	if cfg.MapScaleFactor > 8 {
+		return parseError(
+			"eBPF map scaling factor %d exceeds limit (max: %d)",
+			cfg.MapScaleFactor, MaxArgMapScaleFactor,
+		)
+	}
+
+	if cfg.BpfVerifierLogLevel > 2 {
+		return parseError("invalid eBPF verifier log level: %d", cfg.BpfVerifierLogLevel)
+	}
+
+	if cfg.ProbabilisticInterval < 1*time.Minute || cfg.ProbabilisticInterval > 5*time.Minute {
+		return parseError(
+			"invalid argument for probabilistic-interval: use " +
+				"a duration between 1 and 5 minutes",
+		)
+	}
+
+	if cfg.ProbabilisticThreshold < 1 ||
+		cfg.ProbabilisticThreshold > tracer.ProbabilisticThresholdMax {
+		return parseError(
+			"invalid argument for probabilistic-threshold. Value "+
+				"should be between 1 and %d",
+			tracer.ProbabilisticThresholdMax,
+		)
+	}
+
+	if !cfg.NoKernelVersionCheck {
+		major, minor, patch, err := tracer.GetCurrentKernelVersion()
+		if err != nil {
+			return fmt.Errorf("failed to get kernel version: %v", err)
+		}
+
+		var minMajor, minMinor uint32
+		switch runtime.GOARCH {
+		case "amd64":
+			if cfg.VerboseMode {
+				minMajor, minMinor = 5, 2
+			} else {
+				minMajor, minMinor = 4, 19
+			}
+		case "arm64":
+			// Older ARM64 kernel versions have broken bpf_probe_read.
+			// https://github.com/torvalds/linux/commit/6ae08ae3dea2cfa03dd3665a3c8475c2d429ef47
+			minMajor, minMinor = 5, 5
+		default:
+			return fmt.Errorf("unsupported architecture: %s", runtime.GOARCH)
+		}
+
+		if major < minMajor || (major == minMajor && minor < minMinor) {
+			return fmt.Errorf("host Agent requires kernel version "+
```

Comment:
Shouldn't we explicit here and return a `ErrorWithExitCode` with `exitFailure`?

Reply:
No. If we don't provide any explicit code, `main.go` returns an exitFailure.

Reply:
But a different caller of Validate() would possibly like to check the code and decide about exiting or going on!? Or in other words, what does "no error code explicitly set" currently mean to the caller? It is not immediately obvious to me, maybe I am missing the idea of how this should evolve.

Let me go one step back: To whom is it important whether the example code (main.go) returns two different status codes if this is not even documented?

Reply:
The more I think about it... better just drop the exitCode stuff. cc @christos68k

Reply:
The intent is : if an error is returned, provide a default error code (in the case of a CLI, an exit code 1).
If something specific is provided, use that.

Code relying on Validate may or may not act in relation to the provided code, but that's up to their intended behavior.

Reply:
Dropping exit code entirely would definitely be fine by me.

Reply:
I would keep the `exitCode` in package `main` where we're returning an exit code to the OS but not here, as `controller` package is more generic than just CLI usage. To simplify, I'd get rid of `exitParseError` and only return `exitSuccess` / `exitFailure` from `controller` callers in `main`. This means that we can remove `parseError` from controller and simply return plain errors.